src/document/doc-utils.js

```diff
@@ -180,42 +180,127 @@ function removeLines(doc) {
+  doc &&
+  doc.type === "line" &&
+  doc.hard &&
+  nextDoc &&
+  nextDoc.type === "break-parent";
+function stripDocTrailingHardlineFromDoc(doc) {
+  if (!doc) {
+    return doc;
   }
-  if (lastPart.type === "group") {
-    parts = lastPart.contents.parts;
+
+  switch (doc.type) {
+    case "concat":
+    case "fill": {
+      const { parts } = doc;
+
+      while (parts.length > 1 && isHardline(...parts.slice(-2))) {
+        parts.length -= 2;
+      }
+
+      if (parts.length > 0) {
+        const lastPart = stripDocTrailingHardlineFromDoc(
+          parts[parts.length - 1]
+        );
+        parts[parts.length - 1] = lastPart;
+      }
+      break;
+    }
+    case "align":
+    case "indent":
+    case "group":
+    case "line-suffix":
+      doc.contents = stripDocTrailingHardlineFromDoc(doc.contents);
+      break;
+    case "if-break":
+      doc.breakContents = stripDocTrailingHardlineFromDoc(doc.breakContents);
+      doc.flatContents = stripDocTrailingHardlineFromDoc(doc.flatContents);
+      break;
   }
-  return parts;
+
+  return doc;
 }
 
-function stripTrailingHardline(doc, withInnerParts = false) {
+function stripTrailingHardline(doc) {
   // HACK remove ending hardline, original PR: #1984
-  if (doc.type === "concat" && doc.parts.length !== 0) {
-    const parts = withInnerParts ? getInnerParts(doc) : doc.parts;
-    const lastPart = parts[parts.length - 1];
-    if (lastPart.type === "concat") {
-      if (
-        lastPart.parts.length === 2 &&
-        lastPart.parts[0].hard &&
-        lastPart.parts[1].type === "break-parent"
-      ) {
-        return { type: "concat", parts: parts.slice(0, -1) };
+  return stripDocTrailingHardlineFromDoc(cleanDoc(doc));
+}
+
+const isConcat = (doc) => doc && doc.type === "concat";
+function cleanDocFn(doc) {
+  switch (doc.type) {
+    case "align":
+    case "indent":
+    case "group":
+    case "line-suffix":
+      if (!doc.contents) {
+        return "";
+      }
+      break;
+    case "if-break":
+      if (!doc.flatContents && !doc.breakContents) {
+        return "";
       }
+      break;
+  }
 
-      return {
-        type: "concat",
-        parts: doc.parts.slice(0, -1).concat(stripTrailingHardline(lastPart)),
-      };
+  if (!isConcat(doc)) {
+    return doc;
+  }
+
+  const { parts } = doc;
+  for (let index = parts.length - 1; index >= 0; index--) {
+    let currentDoc = parts[index];
+    // Next doc already normalized
+    const nextDoc = parts[index + 1];
+
+    // Concat continuous string
+    if (typeof currentDoc === "string" && typeof nextDoc === "string") {
+      parts.splice(index + 1, 1);
+      currentDoc += nextDoc;
+    }
+    // Concat continuous `concat`
+    else if (isConcat(currentDoc) && isConcat(nextDoc)) {
+      parts.splice(index + 1, 1);
+      currentDoc.parts.push(...nextDoc.parts);
+      currentDoc = cleanDocFn(currentDoc);
+    }
+    // Concat `concat` and string
+    else if (isConcat(currentDoc) && typeof nextDoc === "string") {
```

Comment:
Can you please add a test for what is achieved with these manipulations? As this function is passed to `mapDoc`, which by itself is recursive and depth-first, I'm not sure about the neccessity of these recursive calls here.

Reply:
Yes, there are recursive calls, but the all needed. Because the inner "concat" already "cleaned", but it's updated, have to clean it again , I can't figure a better way.

Reply:
Let's find a better way to do this. These nested recursions look suspiciously like they're O(n**2).

Reply:
I may found something, I forgot to take "concat.parts" out, will try.